helix-term/src/commands.rs

```diff
@@ -1739,6 +1739,42 @@ mod cmd {
         doc.reload(view.id)
     }
 
+    fn lsp_restart(
+        cx: &mut compositor::Context,
+        _args: &[&str],
+        _event: PromptEvent,
+    ) -> anyhow::Result<()> {
+        let view = cx.editor.tree.get(cx.editor.tree.focus);
+        let doc_id = view.doc;
+        let doc = cx.editor.document(doc_id).unwrap();
+        let language_server = match doc.language_server_arc() {
+            Some(language_server) => language_server,
+            None => return Err(anyhow!("running lsp server not found")),
+        };
+
+        log::debug!("got lspserver");
+        let fut = async move {
+            log::debug!("trying shutdown");
+            let shutdown_result = language_server.shutdown_and_exit().await;
+            if let Err(err) = shutdown_result {
+                log::debug!("shutdown & exit request returned error: {}", err); // we're not doing anything except logging, if LSP crashes then this error is expected
+            }
+            let callback: job::Callback = Box::new(move |editor: &mut Editor, _compositor| {
+                let doc = editor.document(doc_id).unwrap();
+                let language_config = doc.language_config_arc().unwrap();
+                let language_server = editor.language_servers.get(&language_config).unwrap();
```

Comment:
You're getting the failures because `language_servers.get()` returns the same server you just shut down. We maintain one server instance per language because it's shared across files. I think what you really want to do is add a `language_server.restart()` method that will do this inside the server. 

It's hard to get mutable access though, so I imagine we actually want to use arc-swap here to start a new server, arc-swap the new server in. It's a bit trickier than that, you also need to loop over all the documents and re-send textDocument/open events for all open files with that language server, as well as replay any changes since last file save (you can do this by taking the last saved version id, then replaying all the changes in the tree from that id to current id).